File: 2028-the-earliest-and-latest-rounds-where-players-compete/the-earliest-and-latest-rounds-where-players-compete.java
// class Solution {
//     public int[] earliestAndLatest(int n, int firstPlayer, int secondPlayer) {
        
//     }
// }

import java.util.*;

class Solution {
    Map<String, int[]> memo = new HashMap<>();

    public int[] earliestAndLatest(int n, int firstPlayer, int secondPlayer) {
        return dfs(1, n, firstPlayer, secondPlayer);
    }

    private int[] dfs(int l, int r, int fp, int sp) {
        if (fp + sp == l + r) return new int[]{1, 1};
        String key = l + "," + r + "," + fp + "," + sp;
        if (memo.containsKey(key)) return memo.get(key);

        int len = r - l + 1;
        int mid = (len + 1) / 2;

        List<int[]> nextPairs = new ArrayList<>();

        // generate all possible player positions
        List<Integer> players = new ArrayList<>();
        for (int i = l; i <= r; i++) players.add(i);

        // collect all next possible sets based on match outcomes
        getNext(players, 0, players.size() - 1, new ArrayList<>(), nextPairs, fp, sp);

        int minRound = Integer.MAX_VALUE;
        int maxRound = Integer.MIN_VALUE;

        for (int[] next : nextPairs) {
            Arrays.sort(next);
            int nf = -1, ns = -1;
            for (int i = 0; i < next.length; i++) {
                if (next[i] == fp) nf = i + 1;
                if (next[i] == sp) ns = i + 1;
            }
            int[] res = dfs(1, next.length, nf, ns);
            minRound = Math.min(minRound, res[0] + 1);
            maxRound = Math.max(maxRound, res[1] + 1);
        }

        int[] ans = new int[]{minRound, maxRound};
        memo.put(key, ans);
        return ans;
    }

    private void getNext(List<Integer> players, int i, int j, List<Integer> path, List<int[]> res, int fp, int sp) {
        if (i > j) {
            res.add(path.stream().mapToInt(k -> k).toArray());
            return;
        }

        if (i == j) {
            path.add(players.get(i));
            getNext(players, i + 1, j - 1, path, res, fp, sp);
            path.remove(path.size() - 1);
            return;
        }

        int a = players.get(i);
        int b = players.get(j);

        if ((a == fp && b == sp) || (a == sp && b == fp)) return;

        // one of them is fp or sp -> winner is that player
        if (a == fp || b == fp) {
            path.add(fp);
            getNext(players, i + 1, j - 1, path, res, fp, sp);
            path.remove(path.size() - 1);
        } else if (a == sp || b == sp) {
            path.add(sp);
            getNext(players, i + 1, j - 1, path, res, fp, sp);
            path.remove(path.size() - 1);
        } else {
            // both are ordinary -> try both possibilities
            path.add(a);
            getNext(players, i + 1, j - 1, path, res, fp, sp);
            path.remove(path.size() - 1);

            path.add(b);
            getNext(players, i + 1, j - 1, path, res, fp, sp);
            path.remove(path.size() - 1);
        }
    }
}

File: 2028-the-earliest-and-latest-rounds-where-players-compete/the-earliest-and-latest-rounds-where-players-compete.py
from typing import List
from functools import lru_cache

class Solution:
    def earliestAndLatest(self, n: int, firstPlayer: int, secondPlayer: int) -> List[int]:
        @lru_cache(None)
        def dfs(players: tuple) -> tuple[int, int]:
            round_num = 1
            a, b = players.index(firstPlayer), players.index(secondPlayer)
            if a + b == len(players) - 1:
                return (1, 1)
            min_round, max_round = float('inf'), 0
            next_round = []

            def simulate(i: int, j: int, current: List[int]):
                if i > j:
                    next_rounds.append(tuple(sorted(current)))
                    return
                if i == j:
                    current.append(players[i])
                    simulate(i + 1, j - 1, current)
                    current.pop()
                    return
                p1, p2 = players[i], players[j]
                if {p1, p2} == {firstPlayer, secondPlayer}:
                    return
                elif p1 in {firstPlayer, secondPlayer}:
                    current.append(p1)
                    simulate(i + 1, j - 1, current)
                    current.pop()
                elif p2 in {firstPlayer, secondPlayer}:
                    current.append(p2)
                    simulate(i + 1, j - 1, current)
                    current.pop()
                else:
                    current.append(p1)
                    simulate(i + 1, j - 1, current)
                    current.pop()
                    current.append(p2)
                    simulate(i + 1, j - 1, current)
                    current.pop()

            next_rounds = []
            simulate(0, len(players) - 1, [])
            for nxt in next_rounds:
                er, lr = dfs(nxt)
                min_round = min(min_round, er + 1)
                max_round = max(max_round, lr + 1)
            return (min_round, max_round)

        # Initial call with tuple of players in order
        return list(dfs(tuple(range(1, n + 1))))



